src/PhpSpec/Factory/ObjectFactory.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of PhpSpec, A php toolset to drive emergent
+ * design by specification.
+ *
+ * (c) Marcello Duarte <marcello.duarte@gmail.com>
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace PhpSpec\Factory;
+
+use PhpSpec\Exception\Fracture\FactoryDoesNotReturnObjectException;
+
+class ObjectFactory
+{
+    /**
+     * @throws FactoryDoesNotReturnObjectException when callable returns non object
+     */
+    public static function instantiateFromCallable(
+        callable $callable,
+        array $arguments = []
+    ): object {
+        $instance = \call_user_func_array($callable, $arguments);
+
+        if (!\is_object($instance)) {
+            throw new FactoryDoesNotReturnObjectException(sprintf(
+                'The %s did not return an object, returned %s instead',
+                self::callableToString($callable),
+                \gettype($instance)
+            ));
+        }
+
+        return $instance;
+    }
+
+    private static function callableToString(callable $callable): string
```

Comment:
As this takes callable it needs to handle

a) arrays with `[0]` as an instance
b) Closure objects